java/org/apache/catalina/servlets/DefaultServlet.java

```diff
@@ -2611,6 +2574,44 @@ private PrecompressedResource(WebResource resource, CompressionFormat format) {
         }
     }
 
+    /**
+     * RFC 7232 requires weak comparison for If-None-Match
+     */
+    private boolean matchByEtagWeak(String headerValue, String eTag) {
+        // Match W/"1" and W/"1"
+        if (headerValue.contains(eTag)) {
```

Comment:
Yes, it may look not so obvious for those who never used such approach. Still it makes a lot of sense to use this.
In 99% (if not all 100) cases the If-None-Match will be just a single ETag that Tomcat generated itself.
But create an instance of the `StringTokenizer` allocates 160 bytes, and each a new `nextToken()` call takes about 88 bytes for etag like `W/"1047-1578315296666"`.
The contains() call have no any allocations and it is optimized intrinsic function in JVM

Reply:
I *believe* just doing "contains()" will yield incorrect results if the field value is malformed, such as `WW/"1"`.

Reply:
GIGO

Reply:
Garbage in, garbage out is not an acceptable way to handle HTTP headers. Failure to reject malformed HTTP headers from clients can lead to security issues - typically request smuggling - when a proxy takes a different approaching to allowing the invalid header to the back-end server although in this instance the proxy would need to do something pretty unusual.
The right solution here is to implement RFC 7232 compliant parsing of `entity-tag` in `org.apache.tomcat.util.http.parser`
Tomcat has been tightening up the parsing of HTTP headers over time, generally improving things as parsing issues are raised with each header. It appears that now is the time to address `entity-tag`. I also note that we should make the use of weak comparison for `If-Match` configurable. Users that extend the Default servlet to provide strong ETags (or use a custom resource implementation) will almost certainly want to use a strong comparison here.
I'll start work on a suitable parser.